pkg/traceql/engine_metrics.go

```diff
@@ -0,0 +1,565 @@
+		// Invalid
+		return -1
+	}
+
+	return int((ts - start) / step)
+}
+
+type TimeSeries struct {
+	Labels labels.Labels
+	Values []float64
+}
+
+// SeriesSet is a set of unique timeseries. They are mapped by the "Prometheus"-style
+// text description: {x="a",y="b"}
+type SeriesSet map[string]TimeSeries
+
+// VectorAggregator turns a vector of spans into a single numeric scalar
+type VectorAggregator interface {
+	Observe(s Span)
+	Sample() float64
+}
+
+// RangeAggregator sorts spans into time slots
+// TODO - for efficiency we probably combine this with VectorAggregator (see todo about CountOverTimeAggregator)
+type RangeAggregator interface {
+	Observe(s Span)
+	Samples() []float64
+}
+
+// SpanAggregator sorts spans into series
+type SpanAggregator interface {
+	Observe(Span)
+	Series() SeriesSet
+}
+
+// CountOverTimeAggregator counts the number of spans. It can also
+// calculate the rate when given a multiplier.
+// TODO - Rewrite me to be []float64 which is more efficient
+type CountOverTimeAggregator struct {
+	count    float64
+	rateMult float64
+}
+
+var _ VectorAggregator = (*CountOverTimeAggregator)(nil)
+
+func NewCountOverTimeAggregator() *CountOverTimeAggregator {
+	return &CountOverTimeAggregator{
+		rateMult: 1.0,
+	}
+}
+
+func NewRateAggregator(rateMult float64) *CountOverTimeAggregator {
+	return &CountOverTimeAggregator{
+		rateMult: rateMult,
+	}
+}
+
+func (c *CountOverTimeAggregator) Observe(_ Span) {
+	c.count++
+}
+
+func (c *CountOverTimeAggregator) Sample() float64 {
+	return c.count * c.rateMult
+}
+
+// StepAggregator sorts spans into time slots using a step interval like 30s or 1m
+type StepAggregator struct {
+	start   uint64
+	end     uint64
+	step    uint64
+	vectors []VectorAggregator
+}
+
+var _ RangeAggregator = (*StepAggregator)(nil)
+
+func NewStepAggregator(start, end, step uint64, innerAgg func() VectorAggregator) *StepAggregator {
+	intervals := IntervalCount(start, end, step)
+	vectors := make([]VectorAggregator, intervals)
+	for i := range vectors {
+		vectors[i] = innerAgg()
+	}
+
+	return &StepAggregator{
+		start:   start,
+		end:     end,
+		step:    step,
+		vectors: vectors,
+	}
+}
+
+func (s *StepAggregator) Observe(span Span) {
+	interval := IntervalOf(span.StartTimeUnixNanos(), s.start, s.end, s.step)
+	if interval == -1 {
+		return
+	}
+	s.vectors[interval].Observe(span)
+}
+
+func (s *StepAggregator) Samples() []float64 {
+	ss := make([]float64, len(s.vectors))
+	for i, v := range s.vectors {
+		ss[i] = v.Sample()
+	}
+	return ss
+}
+
+const maxGroupBys = 5 // TODO - This isn't ideal but see comment below.
+
+// FastValues is an array of attribute values (static values) that can be used
+// as a map key.  This offers good performance and works with native Go maps and
+// has no chance for collisions (whereas a hash32 has a non-zero chance of
+// collisions).  However it means we have to arbitrarily set an upper limit on
+// the maximum number of values.
+type FastValues [maxGroupBys]Static
```

Comment:
This is used for group by, right? Is the order of `Static` values in `FastValues` relevant?

Reply:
Correct, there is a direct match between the group bys and the values in the array.  `by(foo,bar)` will have values `[ <f>, <b>, nil, nil, nil]`  The surrounding GroupingAggregator has knowledge of both and combines them to create the full time-series at the end:  `{ foo="<f>", bar="<b>"}`